File: design-patterns/class-03.09/payments/paypal-factory.go
package payments

import (
	"errors"
	"strings"
)

type PayPalFactory struct {
	email string
}

func NewPayPalFactory(email string) *PayPalFactory {
	return &PayPalFactory{email: email}
}

func (f *PayPalFactory) CreatePayment() (Payment, error) {
	if !strings.Contains(f.email, "@") {
		return nil, errors.New("email inválido para PayPal")
	}

	return PayPalPayment{email: f.email}, nil
}

File: design-patterns/class-08.10/02.question/PayPalFactory.java
public class PayPalFactory extends PaymentFactory {
  private final String email;

  PayPalFactory(String email) {
    this.email = email;
  }

  @Override
  public Payment createPayment() {
    if (!validarContaPayPal(email)) {
      throw new IllegalArgumentException("E-mail não está vinculado a uma conta PayPal.");
    }

    return new PayPalPayment();
  }

  private boolean validarContaPayPal(String email) {
    // Aqui seria integração real com a API do PayPal.
    // No exemplo, vamos supor que apenas e-mails terminados com "@paypal.com" são válidos.
    return email != null && email.contains("@") && email.endsWith("@paypal.com");
}
}
